pkg/plugin/plugin.go

```diff
@@ -0,0 +1,113 @@
+// and performing executable filename lookups to search
+// for valid plugin files, and execute found plugins.
+type Handler interface {
+	// exists at the given filename, or a boolean false.
+	// Lookup will iterate over a list of given prefixes
+	// in order to recognize valid plugin filenames.
+	// The first filepath to match a prefix is returned.
+	Lookup(filename string) (string, bool)
+	// Execute receives an executable's filepath, a slice
+	// of arguments, and a slice of environment variables
+	// to relay to the executable.
+	Execute(executablePath string, cmdArgs, environment []string) error
+}
+
+// DefaultPluginHandler implements PluginHandler
+type DefaultPluginHandler struct {
+	ValidPrefixes []string
+}
+
+// NewDefaultPluginHandler instantiates the DefaultPluginHandler with a list of
+// given filename prefixes used to identify valid plugin filenames.
+func NewDefaultPluginHandler(validPrefixes []string) *DefaultPluginHandler {
+	return &DefaultPluginHandler{
+		ValidPrefixes: validPrefixes,
+	}
+}
+
+// Lookup implements PluginHandler
+func (h *DefaultPluginHandler) Lookup(filename string) (string, bool) {
+	for _, prefix := range h.ValidPrefixes {
+		path, err := exec.LookPath(fmt.Sprintf("%s-%s", prefix, filename))
+		if err != nil || len(path) == 0 {
+			continue
+		}
+		return path, true
+	}
+
+	return "", false
+}
+
+// Execute implements PluginHandler
+func (h *DefaultPluginHandler) Execute(executablePath string, cmdArgs, environment []string) error {
+
+	// Windows does not support exec syscall.
+	if runtime.GOOS == "windows" {
+		// #nosec G204
+		cmd := exec.Command(executablePath, cmdArgs...)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		cmd.Stdin = os.Stdin
+		cmd.Env = environment
+		err := cmd.Run()
+		if err == nil {
+			os.Exit(0)
+		}
+		return err
+	}
+
+	// invoke cmd binary relaying the environment and args given
+	// append executablePath to cmdArgs, as execve will make first argument the "binary name".
+	// #nosec G204
+	return syscall.Exec(executablePath, append([]string{executablePath}, cmdArgs...), environment)
+}
+
+// HandlePluginCommand receives a pluginHandler and command-line arguments and attempts to find
+// a plugin executable on the PATH that satisfies the given arguments.
+func HandlePluginCommand(pluginHandler Handler, cmdArgs []string) error {
+	remainingArgs := []string{} // all "non-flag" arguments
+
+	for idx := range cmdArgs {
+		if strings.HasPrefix(cmdArgs[idx], "-") {
+			break
+		}
+		remainingArgs = append(remainingArgs, strings.Replace(cmdArgs[idx], "-", "_", -1))
+	}
+
+	foundBinaryPath := ""
+
+	// attempt to find binary, starting at longest possible name with given cmdArgs
+	for len(remainingArgs) > 0 {
+		path, found := pluginHandler.Lookup(strings.Join(remainingArgs, "-"))
+		if !found {
+			remainingArgs = remainingArgs[:len(remainingArgs)-1]
+			continue
+		}
+
+		foundBinaryPath = path
+		break
+	}
+
+	if len(foundBinaryPath) == 0 {
+		return nil
```

Comment:
I think `return nil` has ambiguous meaning with current contract.  It means "not found, use built-in command" here, but it means "executed plugin (on windows), with no issues" below?

Or did you mean to return a "not found" error here?

Reply:
One idea for a (major) refactoring is splitting this function report whether it found a plugin, and let main call execute it separately.  But the catch is that it also has to communicate how many arguments matched plugin name / how many to pass to it...  

One "OOP" way to do this cleanly is make an interface that just has an `Execute() error` method, with no args — an object satisfying it should already carry necessary info (executablePath and args) inside.
Then this function becomes one that either returns such an object, or nil if not found (and maybe an error?).
The object .Execute() can be called separately and that can return an error about execution (or print it and os.Exit).

This is just an idea, and probably overkill.  There are simpler changes that fix it...

But please decide what is the contract of this function for following scenarios:
1. no plugin found
2. plugin found, some OS error starting it
3. plugin found, process finished, non-0 exit status
4. plugin found, process finished, 0 exit status.

(1) and (2) may be same to caller, especially under current policy that built-in command wins over plugin of same name.

Reply:
Hi, see the codes in the main.go also, the general logic here is:
1. Find the command in root cmd tree, if cannot find, then try to find the plugin
2.If the plugin is found, then execute  it, and return what ever error the plugin returned.
3. If the plugin is also not found, we return nil, not error, so the main func will continue to run with the ocm root cmd, and it will return errors cause the cmd not exists.

Reply:
LGTM :+1:  
@igoihman please merge/review.